src/main/java/io/jenkins/plugins/gitlabserver/servers/GitLabServer.java

```diff
@@ -247,34 +223,48 @@ public FormValidation doTestConnection(@QueryParameter String serverUrl,
         /**
          * Stapler form completion.
          *
-         * @param serverUrl the server URL.
+         * @param serverUrl     the server URL.
          * @param credentialsId the credentials Id
          * @return the available credentials.
          */
         @Restricted(NoExternalUse.class) // stapler
         @SuppressWarnings("unused")
-        public ListBoxModel doFillCredentialsIdItems(@QueryParameter String serverUrl, @QueryParameter String credentialsId) {
-            if(!Jenkins.getInstance().hasPermission(Jenkins.ADMINISTER)) {
+        public ListBoxModel doFillCredentialsIdItems(@QueryParameter String serverUrl,
+                                                     @QueryParameter String credentialsId) {
+            if (!Jenkins.getInstance().hasPermission(Jenkins.ADMINISTER)) {
                 return new StandardListBoxModel().includeCurrentValue(credentialsId);
             }
             return new StandardListBoxModel()
-                .includeEmptyValue()
-                .includeMatchingAs(ACL.SYSTEM,
-                        Jenkins.getInstance(),
-                        StandardCredentials.class,
-                        fromUri(serverUrl).build(),
-                        credentials -> credentials instanceof PersonalAccessTokenImpl
-                );
+                    .includeEmptyValue()
+                    .includeMatchingAs(ACL.SYSTEM,
+                            Jenkins.getInstance(),
+                            StandardCredentials.class,
+                            fromUri(serverUrl).build(),
+                            credentials -> credentials instanceof PersonalAccessTokenImpl);
         }
-    }
 
-    /**
-     *  Helper function
-     *
-     * @param gitLabAuth a generic auth object
-     * @return true if gitLabAuth is an object of GitLabAuthToken
-     */
-    private static boolean isToken(GitLabAuth gitLabAuth) {
-        return gitLabAuth instanceof GitLabAuthToken;
+        private static String getToken(String serverUrl, String credentialsId) {
+            String privateToken = UNKNOWN_TOKEN;
+            Jenkins.getInstance().checkPermission(Jenkins.ADMINISTER);
+            StandardCredentials creds = CredentialsMatchers.firstOrNull(
+                    lookupCredentials(
+                            StandardCredentials.class,
+                            Jenkins.getActiveInstance(),
+                            ACL.SYSTEM,
+                            fromUri(defaultIfBlank(serverUrl, GITLAB_SERVER_URL)).build()
+                    ),
+                    CredentialsMatchers.allOf(
+                            AuthenticationTokens.matcher(GitLabAuth.class),
+                            CredentialsMatchers.withId(credentialsId)
+                    )
+            );
+            if (creds != null) {
+                GitLabAuth gitLabAuth = AuthenticationTokens.convert(GitLabAuth.class, creds);
+                if (gitLabAuth instanceof GitLabAuthToken) {
```

Comment:
IIRC there is a CredentialsMatcher where you can fetch by type.

Reply:
Hmm... I think you may actually be using that. Is there a reason lines 262-265 are necessary?

Reply:
Lines 262-265 is to check that it is a `Personal Access Token` and not a `Username/Password` credential as `GitLabAuth` can be of both types. And yes `AuthenticationTokens.matcher` provides a matcher by type.

Reply:
Couldn't you just be more specific in your type in the matcher?

Reply:
Yup, this can be done. I will send a new PR for this. I will raise a ticket,

Reply:
Created a ticket here: https://issues.jenkins-ci.org/browse/JENKINS-57884